### PR TITLE
Generate API docs with Doxygen

### LINKED ISSUE #7814: API docs
It would be good for have API docs generated from comments in the C++ source.

Even though the internal Nix library APIs are unstable and therefore not intended for large scale outside consumption, contributers to Nix itself would still benefit.

Also, in conjunction with stable bindings to other languages, we've discussed stabilizing parts of the C++ interfaces. This would naturally be good for that.

### PR DESCRIPTION
**Pull request by Ericson2314**
# Motivation

The motivation is as stated in issue #7814: even though the the C++ API is internal and unstable, people still want it to be well documented for sake of learning, code review, and other purposes that aren't predicated on it being stable.

Per the brief discussion with the Nix team it is:

1. Not built by default as part of the main derivation
2. Can be built in the dev shell
3. Will be built by CI

Basically, it is done analogous to the coverage CI job, which is similarly a "bonus feature".

# Context

Fixes #7814

Many things were modeled off of #7901. Thanks @roberth!

# Checklist for maintainers

Maintainers: tick if completed or explain if not relevant

 - [x] agreed on idea
 - [x] agreed on implementation strategy
 - [ ] tests, as appropriate
   - functional tests - `tests/**.sh`
   - unit tests - `src/*/tests`
   - integration tests - `tests/nixos/*`
 - [ ] documentation in the manual
 - [ ] code and comments are self-explanatory
 - [ ] commit message explains why the change was made
 - [ ] new feature or incompatible change: updated release notes

### CODE REVIEW

**Comment by abathur** [comment]
This FWIW probably isn't relevant for the use case here, but since nix already leans on sqlite I figure I should mention that doxygen has an optional/experimental sqlite output :)

**Comment by Ericson2314** [comment]
If that is useful for e.g. interactive IDE features, it might indeed be relevant. We are, after all, targeting people that want to work on Nix itself, since this is not part of Nix's public interface.

**Comment by abathur** [comment]
I haven't really looked into doxygen-output-consuming ide tooling. 

It could, but I suspect if it already exists it's probably built on the xml output (it's existed longer, it's default/~stable. 

But it does make it a little easier to build some kinds of tooling.

**Comment by roberth** [comment]
A `devdoc` output would be nice, but an independent derivation _should_ be even nicer, although I don't think we can really pull that off yet without overcomplicating things.

**Comment by Ericson2314** [comment]
Yeah I would do that after your `--enable-tests` PR, @roberth :)

**Comment by fricklerhandwerk** [comment]
Triaged in the Nix team meeting:

We absolutely want this, but don't include it in the manual to avoid making the impression this is somehow a public interface. It's just to improve developer experience.

Decision: Please add a it as a new makefile target and separate derivation output.

<details><summary>Complete discussion</summary>

- @tomberek: inspecting structural changes visually will be much easier than reading header files
- @Ericson2314: there will also be more information for the reviewer
- @fricklerhandwerk: we already agreed we want this
  - @tomberek: the open issue was how to make unmistakably clear that this is not a public interface
      - could simply add a different make target or derivation output, bt not include it in the manual
- @Ericson2314: a lot of comments will have to be updated to follow the doxygen convention, this may be a good beginner project
  - @edolstra: wouldn't want beginners document code they don't understand
      - the only thing worse than missing documentation is incorrect documentation
  - @fricklerhandwerk: disagree on the premise; had multiple pairing and mob sessions digging through the code to understand what it does, and it produced sensible documentation
      - it helps a lot with onboarding beginners and spurs a discussion with maintainers
  - @edolstra: PRs are fine, but they should absolutely not be merged without review by someone knowledgeable with the code
      - the documentation should be authoritatively correct
      - @fricklerhandwerk: yes, obviously

</details>

**Comment by Ericson2314** [comment]
OK I think this is ready now!

**Comment by fricklerhandwerk** [comment]
Example class diagram rendered:

structnix_1_1Installable

**Comment by fricklerhandwerk** [comment]
Discussed in Nix team meeting:

- @edolstra: we want API docs, but not sure about doxygen specifically (it's atrociously ugly)
    - can go for it if no better alternatives available, don't know any
        - (no one knows either)
        - @roberth: https://hdoc.io
    - the real issue is that our source code is not suitable to provide useful doxygen output
        - needs nice class hierarchies, and we don't have that
        - everything ends up in the giant `nix` namespace, and that's not useful organisation
        - may incenetivise refactorings though in order to get closer to that, e.g. of the `util` functions
- @Ericson2314: the subject came up related to the parts with more inheritance structure, in order to provide better overview visually
    - it's already delivering value there
- @edolstra: why is the doxygen config 2700 LOC?
    - @Ericson2314: took the defaults
    - @edolstra: try to avoid specifying unmodified fields, only provide overrides
    - @Ericson2314: will do that if possible
- @edolstra: should add a Hydra build result so the docs can be browsed from Hydra
    - @fricklerhandwerk: could link it from the hacking guide
    - @edolstra: Hydra can redirect to the latest build
    - @Ericson2314: this is downstream of that PR though
- Sphinx has some support, but it turns out it uses Doxygen too!
  - @Ericson2314: Seems like a good follow-up task then.
- agreement: idea approved, fix the config file

File: mk/disable-api-docs.mk
Comment (Ericson2314) [inline mk/disable-api-docs.mk:6]: Just for convenience to give the user a better error, modeled after what @roberth did for the tests when they are disabled.

File: flake.nix
Comment (roberth) [inline flake.nix:544]:
```suggestion
        internal-api-docs =
```

It didn't reflect the intent sufficiently.

File: flake.nix
Comment (roberth) [inline flake.nix:549]:
```suggestion
            pname = "nix-internal-api-docs";
```

File: doc/api/local.mk
Comment (roberth) [inline doc/api/local.mk:7]: Rename `api` to `internal-api` in this file?

**Review by roberth — COMMENTED** [review]
Let's make clear that it's our internal api.

File: mk/disable-api-docs.mk
Comment (roberth) [reply mk/disable-api-docs.mk:6]: We could inline these files into the main Makefile that imports them, but it'd clutter up the main Makefile a bit. I think it's nice to keep that one clean.
I expect these files to be essentially write once and read very very rarely.

**Comment by Ericson2314** [comment]
OK @edolstra, I think I fixed everything brought up in the meeting. It's all yours now!

**Comment by Ericson2314** [comment]
Oops I forgot to note hydra build products as discussed in the meeting. This is done now.

**Review by roberth — APPROVED** [review]
Looks like a good start.

File: doc/internal-api/local.mk
Comment (edolstra) [inline doc/internal-api/local.mk:1]: Why not write 
```
ifeq ($(internal_api_docs), yes)
...
else
.PHONY: internal-api-html
internal-api-html:
	@echo "Internal API docs are disabled. Configure with '--enable-internal-api-docs', or avoid calling 'make internal-api-html'."
	@exit 1
endif
```

That way we don't need `mk/disable-internal-api-docs.mk` and the conditional include in `Makefile`.

File: doc/internal-api/local.mk
Comment (edolstra) [inline doc/internal-api/local.mk:3]: I noticed that this is also generating LaTeX docs in `$doc/share/doc/nix/internal-api/latex`, can that be disabled?

**Review by edolstra — APPROVED** [review]

File: doc/internal-api/local.mk
Comment (Ericson2314) [reply doc/internal-api/local.mk:3]: Sure

File: doc/internal-api/local.mk
Comment (Ericson2314) [reply doc/internal-api/local.mk:5]: Fixed!

**Pull request merged**
